nova_vm/src/ecmascript/builtins/builtin_function.rs

```diff
@@ -60,53 +81,196 @@ impl BuiltinFunctionArgs {
     //    described by behaviour using the provided arguments as the values of the corresponding
     //    parameters specified by behaviour. The new function object has internal slots whose names
     //    are the elements of internalSlotsList, and an [[InitialName]] internal slot.
-    // 6. Set func.[[Prototype]] to prototype.
-    // 7. Set func.[[Extensible]] to true.
-    // 8. Set func.[[Realm]] to realm.
-    // NOTE: Heap data is implicitly attached to the Realm so I don't think
-    //       this matters.
-    let _object = heap.create_object_with_prototype(prototype);
+    let object_index = if let Some(prototype) = args.prototype {
+        // If a prototype is set, then check that it is not the %Function.prototype%
+        let realm_function_prototype = agent.get_realm(realm).intrinsics().function_prototype();
+        if prototype == realm_function_prototype {
+            // If the prototype matched the realm function prototype, then ignore it
+            // as the BuiltinFunctionHeapData indirectly implies this prototype.
+            None
+        } else {
+            // If some other prototype is defined then we need to create a backing object.
+            // 6. Set func.[[Prototype]] to prototype.
+            // 7. Set func.[[Extensible]] to true.
+            Some(agent.heap.create_object_with_prototype(prototype))
+        }
+    } else {
+        None
+    };
 
     // 9. Set func.[[InitialName]] to null.
-    // TODO: This is non-standard.
-    let initial_name = heap.create(args.name).into_value();
-    // 10. Perform SetFunctionLength(func, length).
-    let length = args.length as u8;
-    // TODO: Actually set behaviour somewhere
-    let func = heap.create_function(initial_name, length, false);
-
-    // TODO: Steps 11-12
-    // 11. If prefix is not present, then
-    //     a. Perform SetFunctionName(func, name).
-    // 12. Else,
-    //     a. Perform SetFunctionName(func, name, prefix).
+    // Note: SetFunctionName inlined here: We know name is a string
+    let initial_name = if let Some(prefix) = args.prefix {
+        // 12. Else,
+        // a. Perform SetFunctionName(func, name, prefix).
+        String::from_str(agent, &format!("{} {}", args.name, prefix))
+    } else {
+        // 11. If prefix is not present, then
+        // a. Perform SetFunctionName(func, name).
+        String::from_str(agent, args.name)
+    };
+    agent
+        .heap
+        .builtin_functions
+        .push(Some(BuiltinFunctionHeapData {
+            behaviour,
+            initial_name: Some(initial_name),
+            name: Some(initial_name),
+            // 10. Perform SetFunctionLength(func, length).
+            length: args.length as u8,
+            // 8. Set func.[[Realm]] to realm.
+            realm,
+            object_index,
+        }));
 
     // 13. Return func.
-    Function::from(func)
+    BuiltinFunctionIndex::last(&agent.heap.builtin_functions)
```

Comment:
Is there any way this could be a heap `create()` method?

Reply:
Yeah, should be just a matter of changing the call. Good point.